examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java

```diff
@@ -0,0 +1,150 @@
+        // stream's flow control using the response stream's observer, but this is the way it is.
+        final ServerCallStreamObserver<HelloReply> serverCallStreamObserver =
+            (ServerCallStreamObserver<HelloReply>) responseObserver;
+        serverCallStreamObserver.disableAutoInboundFlowControl();
+        // Signal the request sender to send one message.
+        serverCallStreamObserver.request(1);
+
+        // Set up a back-pressure-aware producer for the response stream. The onReadyHandler will be invoked
+        // when the consuming side has enough buffer space to receive more messages.
+        //
+        // Note: the onReadyHandler is invoked by gRPC's internal thread pool. You can't block in this in
+        // method or deadlocks can occur.
+        serverCallStreamObserver.setOnReadyHandler(new Runnable() {
+          @Override
+          public void run() {
+            // Start generating values from where we left off on a non-gRPC thread.
+            pool.execute(new Runnable() {
+              @Override
+              public void run() {
+                // requestStream.isReady() will go false when the consuming side runs out of buffer space and
+                // signals to slow down with back-pressure.
+                while (serverCallStreamObserver.isReady() && !serverCallStreamObserver.isCancelled()) {
+                  if (!work.isEmpty()) {
+                    // Send more messages if there are more messages to send.
+                    String name = work.remove();
+                    if (name != COMPLETED) {
+                      // Simulate doing some work to generate the next value
+                      try {
+                        Thread.sleep(200);
+                      } catch (InterruptedException e) {
+                        e.printStackTrace();
+                      }
+
+                      String message = "Hello " + name;
+                      System.out.println("<-- " + message);
+                      HelloReply reply = HelloReply.newBuilder().setMessage(message).build();
+
+                      // Send a response.
+                      responseObserver.onNext(reply);
+                    } else {
+                      System.out.println("Done.");
+                      responseObserver.onCompleted();
+                    }
+                  }
+                }
+              }
+            });
+          }
+        });
+
+        // Give gRPC a StreamObserver it can write incoming requests into.
+        return new StreamObserver<HelloRequest>() {
+          @Override
+          public void onNext(HelloRequest request) {
+            // Process the request and send a response or an error.
+            try {
+              // Accept and enqueue the request.
+              String name = request.getName();
+              System.out.println("--> " + name);
+              work.add(name);
```

Comment:
The `onReadyHandler` can be triggered while the previous handler is
still running. This is because the handler is called every time the
stream transitions from notReady->ready. The code right now has the
potential to kick off multiple sender threads by mistake. One way to
fix this could be to have only a single sender thread that blocks
rather than returns, and use the `onReadyHandler` to wake the thread
up.

It's possible to make this all work, but the scope of this example is
getting a bit larger than the original goal showing how to use
`request`. Perhaps we can simplify the demo to avoid simultaneous
reading and writing:

 - The server receives the message and responds in the same
   thread. The server can still sleep to simulate doing some work.
 - The client sends requests using logic similar to
   `StreamObservers#copyWithFlowControl` so that it will get pushed
   back.